### PR TITLE
Feature: Deep Neural Operator

### PR DESCRIPTION
**Pull request by JakobEliasWagner**
# Feature: Deep Neural Operator

## Description

Adds a novel operator. The deep neural operator modifies the inputs $x$ and $u$ by using all dimensions of all sensors and sensor locations as input to a linear layer. The architecture utilizes a deep residual network to process the stacked $y$, flat $x$, and flat $u$ efficiently and independent of the size of $y$.

### How does it solve the problem?

  - Implements Deep Neural Operator.

### How are the changes tested?

  - Added unit tests for shapes and integration test (convergence study).

## Checklist for Contributors

- [x] Scope: This PR tackles exactly one problem.
- [x] Conventions: The branch follows the `feature/title-slug` convention.
- [x] Conventions: The PR title follows the `Bugfix: Title` convention.
- [x] Coding style: The code passes all pre-commit hooks.
- [x] Documentation: All changes are well-documented.
- [x] Tests: New features are tested and all tests pass successfully.
- [x] Changelog: Updated CHANGELOG.md for new features or breaking changes.
- [x] Review: A suitable reviewer has been assigned.

## Checklist for Reviewers:

- [ ] The PR solves the issue it claims to solve and only this one.
- [ ] Changes are tested sufficiently and all tests pass.
- [ ] Documentation is complete and well-written.
- [ ] Changelog has been updated, if necessary.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,7 @@
 - Add `timeseries.ipynb` example.
 - Add `Function`, `FunctionSet`, and `FunctionOperatorDataset` classes.
 - Add `function.ipynb` example.
+- Implement `FusionOperator`.
```
Comment (samuelburbulla) [inline CHANGELOG.md:16]:
```suggestion
- Implement `DeepNeuralOperator`.
```

File: src/continuity/operators/deep_neural_operator.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+"""
+`continuity.operators.deep_neural_operator`
+
+The Deep Neural Operator architecture.
```
Comment (samuelburbulla) [inline src/continuity/operators/deep_neural_operator.py:4]:
```suggestion
The Deep Neural Operator (DNO) architecture.
```

File: src/continuity/operators/deep_neural_operator.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import torch
+from continuity.operators import Operator
+from continuity.operators.common import DeepResidualNetwork
+from continuity.data import DatasetShapes
+
+
+class DeepNeuralOperator(Operator):
+    """
+    The `DeepNeuralOperator` class integrates a deep residual network within a neural operator framework. It uses all
+    scalar values of the input locations, input functions, and individual evaluation points as inputs for a deep
```
Comment (samuelburbulla) [inline src/continuity/operators/deep_neural_operator.py:16]:
```suggestion
   input locations, input values, and the evaluation point as input for a deep
```

File: src/continuity/operators/deep_neural_operator.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    residual network.
+
+    Args:
+        shapes: An instance of `DatasetShapes`.
+        width: The width of the Deep Residual Network, defining the number of neurons in each hidden layer.
```
Comment (samuelburbulla) [inline src/continuity/operators/deep_neural_operator.py:21]:
```suggestion
        width: The width of the `DeepResidualNetwork`.
```

File: src/continuity/operators/deep_neural_operator.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        depth: The depth of the Deep Residual Network, indicating the number of hidden layers in the network.
```
Comment (samuelburbulla) [inline src/continuity/operators/deep_neural_operator.py:22]:
```suggestion
        depth: The depth of the `DeepResidualNetwork`.
```

File: src/continuity/operators/deep_neural_operator.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, shapes: DatasetShapes, width: int = 32, depth: int = 3):
+        super().__init__()
+        self.shapes = shapes
+
+        self.width = width
+        self.depth = depth
+
+        self.net = DeepResidualNetwork(
+            input_size=(
+                shapes.y.dim + shapes.u.dim * shapes.u.num + shapes.x.dim * shapes.x.num
```
Comment (samuelburbulla) [inline src/continuity/operators/deep_neural_operator.py:35]:
```suggestion
                shapes.x.dim * shapes.x.num + shapes.u.dim * shapes.u.num + shapes.y.dim
```

File: src/continuity/operators/deep_neural_operator.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            ),
+            output_size=shapes.v.dim,
+            width=width,
+            depth=depth,
+        )
```
Comment (samuelburbulla) [inline src/continuity/operators/deep_neural_operator.py:40]: activation function?

File: tests/operators/test_deep_neural_operator.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        assert output.shape == v.shape
+
+
+@pytest.mark.slow
+def test_does_converge():
+    # Parameters
+    num_sensors = 16
+
+    # Data set
+    dataset = Sine(num_sensors, size=1)
```
Comment (samuelburbulla) [inline tests/operators/test_deep_neural_operator.py:62]: rebase on main

File: src/continuity/operators/deep_neural_operator.py (hunk 3 above)
Comment (JakobEliasWagner) [reply src/continuity/operators/deep_neural_operator.py:16]: Much clearer, thank you :)

**Review by samuelburbulla — APPROVED** [review]

**Pull request merged**